api/v1alpha1/ratelimitfilter_types.go

```diff
@@ -0,0 +1,164 @@
+// +kubebuilder:validation:Enum=Global
+type RateLimitType string
+
+const (
+	// GlobalRateLimitType allows the rate limits to be applied across all Envoy proxy instances.
+	GlobalRateLimitType RateLimitType = "Global"
+)
+
+// GlobalRateLimit defines the global rate limit configuration.
+type GlobalRateLimit struct {
+	// Rules are a list of RateLimit selectors and limits.
+	// Each rule and its associated limit is applied
+	// in a mutually exclusive way i.e. if multiple
+	// rules get selected, each of their associated
+	// limits get applied, so a single traffic request
+	// might increase the rate limit counters for multiple
+	// rules if selected.
+	//
+	// +kubebuilder:validation:MaxItems=16
+	Rules []RateLimitRule `json:"rules"`
+}
+
+// RateLimitRule defines the semantics for matching attributes
+// from the incoming requests, and setting limits for them.
+type RateLimitRule struct {
+	// ClientSelectors holds the list of select conditions to select
+	// specific clients using attributes from the traffic flow.
+	// All individual select conditions must hold True for this rule
+	// and its limit to be applied.
+	// If this field is empty, it is equivalent to True, and
+	// the limit is applied.
+	//
+	// +optional
+	// +kubebuilder:validation:MaxItems=8
+	ClientSelectors []RateLimitSelectCondition `json:"clientSelectors,omitempty"`
+	// Limit holds the rate limit values.
+	// This limit is applied for traffic flows when the selectors
+	// compute to True, causing the request to be counted towards the limit.
+	// The limit is enforced and the request is ratelimited, i.e. a response with
+	// 429 HTTP status code is sent back to the client when
+	// the selected requests have reached the limit.
+	Limit RateLimitValue `json:"limit"`
+}
+
+// RateLimitSelectCondition specifies the attributes within the traffic flow that can
+// be used to select a subset of clients to be ratelimited.
+// All the individual conditions must hold True for the overall condition to hold True.
+type RateLimitSelectCondition struct {
+	// Headers is a list of all header matches that must be matched
+	// for the overall select condition to hold True.
+	//
+	// +listType=map
+	// +listMapKey=name
+	// +optional
+	// +kubebuilder:validation:MaxItems=16
+	Headers []HeaderMatch `json:"headers,omitempty"`
+}
+
+// HeaderMatch defines the match attributes within the HTTP Headers of the request.
+type HeaderMatch struct {
+	// Type specifies how to match against the value of the header.
+	//
+	// +optional
+	// +kubebuilder:default=Exact
+	Type *HeaderMatchType `json:"type,omitempty"`
+
+	// Name of the HTTP header.
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:MaxLength=256
+	Name string `json:"name"`
+
+	// Value within the HTTP header. Due to the
+	// case-insensitivity of header names, "foo" and "Foo" are considered equivalent.
+	// Do not set this field when Type="Distinct", implying matching on any/all unique values within the header.
```

Comment:
This comment can be removed if we drop the `Distinct` type. See https://github.com/envoyproxy/gateway/pull/706/files#r1054657720 for additional details.

Reply:
Two of the three header matching types proposed in this PR are the same as upstream Gateway API. Do you know why Gateway API doesn't support the `Distinct` type? If not, it could be valuable to share our use case and get feedback.

Reply:
its naming as well as use case has already been upvoted by reviewers, so I feel strongly to keep it, and proactively present it to the users, and get feedback for this `v1alpha1` version.

Reply:
> Do you know why Gateway API doesn't support the Distinct type?

"Distinct" isn't really a matching mechanism per se, but rather denotes that each unique value for a header the RLS sees will get its own rate limiting bucket, which doesn't really fit into the idea of matching request attributes to funnel to a backend